File: daemon/src/ws/handlers/chatOpen.ts
/**
 * JSON agent-chat WS handlers (Decision 6/12, pagination R2.1/R2.3/R2.7).
 *
 * `chat:open` subscribes the connection to a session's normalized event stream
 * and replays a BOUNDED window via `chat:replay` — the last N turns + a
 * `{ oldestSeq, hasMore }` keyset cursor — not the whole transcript. When a
 * `sinceSeq` is supplied (reconnect), it instead replays only the delta of
 * events newer than that cursor. It then bridges the session's `JsonAgentStream`
 * → `session:message` / `session:meta` frames for as long as the chat is open.
 *
 * Read→subscribe gap (R2.7): the live listeners are attached BEFORE the snapshot
 * read below, which is synchronous — so no event emitted between the snapshot and
 * the attach can be lost. Any overlap dedupes by `id`/`logSeq` on the client.
 *
 * `chat:close` detaches those listeners. The session's `JsonAgentStream` lives
 * on the (lazily-created) `JsonAgentSession`, so opening a chat before any turn
 * has run still works — the stream just stays quiet until a turn is enqueued.
 */

import type { WSConnection, ChatStreamEntry } from "../connection.js";
import type { ClientMessage } from "../protocol.js";
import {
  findJsonSessionContext,
  resolveJsonAgent,
  readSessionTail,
  readSessionSince,
  type JsonSessionContext,
} from "../../services/jsonAgentChat.js";
import type { NormalizedEvent, SessionMeta } from "../../types.js";

/** Default bounded replay window on open (turns). Open decision #1: N=20. */
const TAIL_TURNS = 20;

export async function handleChatOpen(
  conn: WSConnection,
  msg: Extract<ClientMessage, { type: "chat:open" }>,
): Promise<void> {
  const { sessionId, sinceSeq } = msg;

  const ctx = findJsonSessionContext(sessionId);
  if (!ctx) {
    conn.send({ type: "session:error", sessionId, message: `Session '${sessionId}' not found` });
    return;
  }

  conn.subscribe([sessionId]);

  // Resolve (lazily create) the JsonAgentSession so we can attach to its stream
  // and read its current meta. No spawn happens here — only on enqueue.
  const resolved = await resolveJsonAgent(sessionId, 0).catch(() => null);

  if (!resolved || !resolved.ok) {
    // Not a JSON session or mode unresolved — no live stream to attach; send a
    // bounded snapshot from disk (tail or since delta) and stop.
    sendSnapshot(conn, ctx, sessionId, sinceSeq);
    return;
  }
  const { agent } = resolved;

  // Attach the live listeners FIRST, then take the snapshot synchronously below
  // (R2.7 read→subscribe gap). Replace any prior subscription for this session.
  conn.unregisterChatStream(sessionId);
  const onMessage = (event: NormalizedEvent): void => {
    conn.send({ type: "session:message", sessionId, event });
  };
  const onMeta = (meta: SessionMeta): void => {
    conn.send({ type: "session:meta", sessionId, meta });
  };
  agent.stream.on("message", onMessage);
  agent.stream.on("meta", onMeta);
  const entry: ChatStreamEntry = { stream: agent.stream, onMessage, onMeta };
  conn.registerChatStream(sessionId, entry);

  // Snapshot AFTER attach — synchronous, so nothing interleaves (R2.7).
  sendSnapshot(conn, ctx, sessionId, sinceSeq);
  conn.send({ type: "session:meta", sessionId, meta: agent.getMeta() });
}

/**
 * Send the bounded `chat:replay`: a `sinceSeq` delta (reconnect) or the tail-N
 * window + `{ oldestSeq, hasMore }` cursor (fresh open). Reads live-or-disk.
 */
function sendSnapshot(
  conn: WSConnection,
  ctx: JsonSessionContext,
  sessionId: string,
  sinceSeq: number | undefined,
): void {
  if (sinceSeq !== undefined) {
    conn.send({ type: "chat:replay", sessionId, events: readSessionSince(ctx, sinceSeq) });
    return;
  }
  const page = readSessionTail(ctx, TAIL_TURNS);
  conn.send({
    type: "chat:replay",
    sessionId,
    events: page.events,
    ...(page.oldestSeq !== undefined ? { oldestSeq: page.oldestSeq } : {}),
    hasMore: page.hasMore,
  });
}

export function handleChatClose(
  conn: WSConnection,
  msg: Extract<ClientMessage, { type: "chat:close" }>,
): void {
  conn.unregisterChatStream(msg.sessionId);
  conn.unsubscribe([msg.sessionId]);
}

File: rust/vst-ws/src/handlers/chat_open.rs
//! `chat:open` / `chat:close` handlers — JSON agent-chat WS subscription +
//! bounded replay.
//!
//! Ports `daemon/src/ws/handlers/chatOpen.ts`. `chat:open` subscribes the
//! connection to a session's normalized event stream and replays a BOUNDED
//! window via `chat:replay` (tail-N + keyset cursor, or a `sinceSeq` delta on
//! reconnect). It bridges the session's `JsonAgentStream` → `session:message` /
//! `session:meta` frames for as long as the chat is open.
//!
//! Reaches plugin resolution ONLY through `vst_agents::json_agent_chat` —
//! never imports `vst-agents::registry` directly (System Boundary restricted
//! call-site rule).

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use vst_agents::json_agent_chat::{
    find_json_session_context, read_session_meta, read_session_since, read_session_tail,
    resolve_json_agent, JsonSessionContext,
};
use vst_agents::json_agent_registry::JsonAgentRegistry;
use vst_agents::json_agent_session::JsonAgentSession;
use vst_store::StoreHandle;
use vst_types::domain::LifecycleState;
use vst_types::events::Broadcaster;
use vst_types::ws::ClientMessage;
use vst_types::ws::ServerMessage;
use vst_types::{NormalizedEvent, SessionMeta};

use crate::connection::{ChatStreamEntry, WsConnection};

/// Default bounded replay window on open (turns).
const TAIL_TURNS: i64 = 15;

pub async fn handle_chat_open(
    conn: &WsConnection,
    store: &StoreHandle,
    registry: &JsonAgentRegistry<JsonAgentSession>,
    broadcaster: Broadcaster,
    daemon_port: u16,
    msg: &ClientMessage,
) {
    let ClientMessage::ChatOpen {
        session_id,
        since_seq,
    } = msg
    else {
        return;
    };
    let session_id = session_id.clone();
    let since_seq = *since_seq;

    let Some(ctx) = find_json_session_context(store, &session_id).await else {
        conn.send(ServerMessage::SessionError {
            session_id: session_id.clone(),
            message: format!("Session '{session_id}' not found"),
            reason: None,
        });
        return;
    };

    conn.subscribe(std::slice::from_ref(&session_id));

    // A session marked `done` has had its JsonAgentSession released — resolving
    // it would lazily RE-CREATE it. Serve it from disk instead.
    if ctx.session.lifecycle.state == LifecycleState::Done {
        conn.unregister_chat_stream(&session_id);
        send_snapshot(conn, &ctx, registry, &session_id, since_seq);
        let meta = read_session_meta(&ctx, registry).await;
        conn.send(ServerMessage::SessionMetaEvent {
            session_id: session_id.clone(),
            meta: Box::new(meta),
        });
        return;
    }

    let resolved = resolve_json_agent(&session_id, daemon_port, store, registry, broadcaster)
        .await
        .ok();

    let Some(resolved) = resolved else {
        send_snapshot(conn, &ctx, registry, &session_id, since_seq);
        return;
    };

    // Attach the live listeners FIRST, then take the snapshot synchronously
    // (R2.7 read→subscribe gap). Replace any prior subscription.
    conn.unregister_chat_stream(&session_id);
    let active = Arc::new(AtomicBool::new(true));

    // The fan-out body lives in an `Arc` so it can be shared between the
    // stream listener (registered below) and the stored entry, without
    // registering the same behaviour twice.
    let c = conn.clone();
    let sid = session_id.clone();
    let active_msg = active.clone();
    let on_message_arc: Arc<dyn Fn(&NormalizedEvent) + Send + Sync> = Arc::new(move |event| {
        if active_msg.load(Ordering::SeqCst) {
            c.send(ServerMessage::SessionMessage {
                session_id: sid.clone(),
                event: Box::new(event.clone()),
            });
        }
    });
    let c = conn.clone();
    let sid = session_id.clone();
    let active_meta = active.clone();
    let on_meta_arc: Arc<dyn Fn(&SessionMeta) + Send + Sync> = Arc::new(move |meta| {
        if active_meta.load(Ordering::SeqCst) {
            c.send(ServerMessage::SessionMetaEvent {
                session_id: sid.clone(),
                meta: Box::new(meta.clone()),
            });
        }
    });

    // Register the closures on the stream, and keep the SAME Arc bodies in the
    // entry so `unregister` can mark them inert via the shared `active` flag.
    {
        let f = on_message_arc.clone();
        resolved.agent.stream().on_message(Box::new(move |e| f(e)));
        let g = on_meta_arc.clone();
        resolved.agent.stream().on_meta(Box::new(move |m| g(m)));
    }

    let entry = ChatStreamEntry {
        on_message: Box::new(move |e| on_message_arc(e)),
        on_meta: Box::new(move |m| on_meta_arc(m)),
        active,
    };
    conn.register_chat_stream(&session_id, entry);

    // Snapshot AFTER attach — synchronous, so nothing interleaves (R2.7).
    send_snapshot(conn, &ctx, registry, &session_id, since_seq);
    let meta = resolved.agent.get_meta();
    conn.send(ServerMessage::SessionMetaEvent {
        session_id: session_id.clone(),
        meta: Box::new(meta),
    });
}

fn send_snapshot(
    conn: &WsConnection,
    ctx: &JsonSessionContext,
    registry: &JsonAgentRegistry<JsonAgentSession>,
    session_id: &str,
    since_seq: Option<i64>,
) {
    if let Some(since) = since_seq {
        let page = read_session_since(ctx, registry, since);
        conn.send(ServerMessage::ChatReplay {
            session_id: session_id.to_string(),
            events: page.events,
            oldest_seq: None,
            has_more: Some(page.has_more),
            next_seq: page.next_seq,
        });
    } else {
        let page = read_session_tail(ctx, registry, TAIL_TURNS);
        conn.send(ServerMessage::ChatReplay {
            session_id: session_id.to_string(),
            events: page.events,
            oldest_seq: page.oldest_seq,
            has_more: Some(page.has_more),
            next_seq: None,
        });
    }
}

pub fn handle_chat_close(conn: &WsConnection, msg: &ClientMessage) {
    if let ClientMessage::ChatClose { session_id } = msg {
        conn.unregister_chat_stream(session_id);
        conn.unsubscribe(std::slice::from_ref(session_id));
    }
}
